File: 0101-0200/0150.Evaluate_Reverse_Polish_Notation.js
//4ms





/**
 * @param {string[]} tokens
 * @return {number}
 */
var evalRPN = function(tokens) {
    let stack = [];

    for (let c of tokens) {
        if (c === "+") {
            stack.push(stack.pop() + stack.pop());
        } else if (c === "-") {
            let second = stack.pop();
            let first = stack.pop();
            stack.push(first - second);
        } else if (c === "*") {
            stack.push(stack.pop() * stack.pop());
        } else if (c === "/") {
            let second = stack.pop();
            let first = stack.pop();
            stack.push(parseInt(first / second));
        } else {
            stack.push(parseInt(c));
        }
    }

    return stack[0];    
};






//8ms







/**
 * @param {string[]} tokens
 * @return {number}
 */
var evalRPN = function(tokens) {
    const stack=[]
    tokens.forEach(token=>{
        if (token=='+') {
            stack.push(stack.pop()+stack.pop())
        }
        else if (token=='*') {
            stack.push(stack.pop()*stack.pop())
        }    
        else if (token=='-') {
            const num1=stack.pop()
            const num2=stack.pop()
            stack.push(num2-num1)
        }   
        else if (token=='/') {
            const num1=stack.pop()
            const num2=stack.pop()
            const sign = num1*num2>0?1:-1
            stack.push(Math.floor(Math.abs(num2)/Math.abs(num1))*sign)
        } 
        else {
            stack.push(parseInt(token))
        }   
        //console.log(stack[stack.length-1])
    })
    return stack.pop()
};

File: 0101-0200/0150.Evaluate_Reverse_Polish_Notation.java
//5ms





class Solution {
    public int evalRPN(String[] tokens) {
        Stack<Integer> stack = new Stack<>();

        for (String c : tokens) {
            if (c.equals("+")) {
                stack.push(stack.pop() + stack.pop());
            } else if (c.equals("-")) {
                int second = stack.pop();
                int first = stack.pop();
                stack.push(first - second);
            } else if (c.equals("*")) {
                stack.push(stack.pop() * stack.pop());
            } else if (c.equals("/")) {
                int second = stack.pop();
                int first = stack.pop();
                stack.push(first / second);
            } else {
                stack.push(Integer.parseInt(c));
            }
        }

        return stack.peek();        
    }
}





//0ms






class Solution {
    int i;

    public int evalRPN(String[] tokens) {
        i = tokens.length; // traversing from back side
        return eval(tokens);
    }

    public int eval(String[] tokens) {
        String currentString = tokens[--i];
        int k = 0, num = 0, sign = 1;
        char s = currentString.charAt(0);
        
        if (currentString.length() == 1) {
            switch (s) {
                case '+':
                    return eval(tokens) + eval(tokens);
                    //no break is used since its a return statement
                case '-':
                    return -eval(tokens) + eval(tokens);
                case '*':
                    return eval(tokens) * eval(tokens);
                case '/':
                    int second = eval(tokens);
                    int first = eval(tokens);
                    return first / second;
                default:
                    return s - '0'; // convert string to integer
            }
        }
        
        else {
            if (s == '-') {
                sign = -1;
                k++;
            }
            while (k < currentString.length()) {
                s = currentString.charAt(k++);
                num = num * 10 + s - '0';
            }
            return num * sign;
        }
    }
}
